mypy/server/update.py

```diff
@@ -361,6 +367,45 @@ def update_module(self,
         return remaining, (module, path), None
 
 
+def find_unloaded_deps(manager: BuildManager, graph: Dict[str, State],
+                       initial: Sequence[str]) -> List[str]:
+    """Find all the deps of the nodes in initial that haven't had their tree loaded.
+
+    The key invariant here is that if a module is loaded, so are all
+    of their dependencies. This means that when we encounter a loaded
+    module, we don't need to explore its dependencies.  (This
+    invariant is slightly violated when dependencies are added, which
+    can be handled by calling find_unloaded_deps directly on the new
+    dependencies)
+    """
+    worklist = list(initial)
+    seen = set()  # type: Set[str]
+    unloaded = []
+    while worklist:
+        node = worklist.pop()
+        if node in seen or node not in graph:
+            continue
+        seen.add(node)
+        if node not in manager.modules:
+            continue
+        if isinstance(manager.modules[node], UnloadedMypyFile):
+            ancestors = graph[node].ancestors or []
+            worklist.extend(graph[node].dependencies + ancestors)
+            unloaded.append(node)
+
+    return unloaded
+
+
+def ensure_trees_loaded(manager: BuildManager, graph: Dict[str, State],
+                        initial: Sequence[str]) -> None:
+    """Ensure that the modules in initial and their deps have loaded trees"""
+    to_process = find_unloaded_deps(manager, graph, initial)
+    if to_process:
+        manager.log("Calling process_fresh_scc on an 'scc' of size {} ({})".format(
+            len(to_process), to_process))
+        process_fresh_scc(graph, to_process, manager)
```

Comment:
The modules to process may contain multiple SCCs, so the `process_fresh_scc` function name is a misnomer. Maybe rename the function, and mention in the docstring that it can be used to process a SCC?